source/android/logging.c

```diff
@@ -0,0 +1,187 @@
+        if (thread_id_written < 0) {
+            return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        }
+        current_index = s_advance_and_clamp_index(current_index, thread_id_written, fake_total_length);
+    }
+
+    if (current_index < fake_total_length) {
+        /* output subject name */
+        if (formatting_data->subject) {
+            int subject_written = snprintf(
+                formatting_data->buffer + current_index,
+                fake_total_length - current_index,
+                "[%s]",
+                formatting_data->subject);
+
+            if (subject_written < 0) {
+                return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+            }
+
+            current_index = s_advance_and_clamp_index(current_index, subject_written, fake_total_length);
+        }
+    }
+
+    if (current_index < fake_total_length) {
+        int separator_written =
+            snprintf(formatting_data->buffer + current_index, fake_total_length - current_index, " - ");
+        current_index = s_advance_and_clamp_index(current_index, separator_written, fake_total_length);
+    }
+
+    if (current_index < fake_total_length) {
+        /*
+         * Now write the actual data requested by the user
+         */
+
+        int written_count = vsnprintf(
+            formatting_data->buffer + current_index, fake_total_length - current_index, formatting_data->format, args);
+
+        if (written_count < 0) {
+            return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        }
+
+        current_index = s_advance_and_clamp_index(current_index, written_count, fake_total_length);
+    }
+
+    /*
+     * End with a newline.
+     */
+    int newline_written_count =
+        snprintf(formatting_data->buffer + current_index, formatting_data->total_length - current_index, "\n");
+    if (newline_written_count < 0) {
+        return aws_raise_error(AWS_ERROR_UNKNOWN); /* we saved space, so this would be crazy */
+    }
+
+    formatting_data->bytes_written = current_index + newline_written_count;
+
+    return AWS_OP_SUCCESS;
+}
+
+static struct aws_logger_logcat { enum aws_log_level level; } s_logcat_impl;
+
+static int s_logcat_log(
+    struct aws_logger *logger,
+    enum aws_log_level log_level,
+    aws_log_subject_t subject,
+    const char *format,
+    ...) {
+    (void)logger;
+
+    va_list format_args;
+    va_start(format_args, format);
+
+    char buffer[LOGCAT_MAX_BUFFER_SIZE];
+    struct logcat_format_data fmt = {
+        .buffer = buffer,
+        .total_length = AWS_ARRAY_SIZE(buffer),
+        .format = format,
+        .subject = aws_log_subject_name(subject),
+    };
+
+    int result = s_logcat_format(&fmt, format_args);
+
+    va_end(format_args);
+
+    if (result != AWS_OP_SUCCESS) {
+        return AWS_OP_ERR;
+    }
+
+    /* ANDROID_LOG_VERBOSE = 2, ANDROID_LOG_FATAL = 7 */
+    const int prio = 0x8 - log_level;
+    __android_log_write(prio, "AWSCRT", buffer);
```

Comment:
Seems more useful to send the actual subject to logcat, instead of "AWSCRT" for everything.

I've always dreamed of filtering different topics at different levels and that's built into logcat. You prevent people from doing that by classifying everything as "AWSCRT"

Reply:
I will fix this in a follow up